### DISCUSSION TITLE
How to use Turborepo with Next.js 13 without "transpilePackages"

### DISCUSSION THREAD

**Discussion by heloineto** [vercel/turborepo#5813] [Help] [answered]
Hi. I have a monorepo with Vite projects and react libraries with components and custom hooks. I am using `tsup` to compile the packages (same setup as the official Vite example: https://github.com/vercel/turbo/tree/main/examples/with-vite)

I want to add a new Next.js 13 project using the `app` directory.

However, I was overwhelmed with all kinds of errors. I added all the `"use client";` directives necessary on clients components, but still no luck.

I believe one of the problems is that tsup compiles everything to a single file, so the "use client" directives don't work properly.

Also, I noticed that the official Next.js 13 example does not use a build step on the react library and uses the "transpilePackages" option on the `next.config.js` file. This is not what I want at all. I want to develop packages that I can use across Next.js, Vite, CRA and possibly submit to npm.

Does anyone have a working example of a turborepo with a Next.js 13 app and a Vite app sharing a react library? How can this be achieved?

**Comment by sheeehy** [comment]
Its hard to find examples using Next.js 13 as its still pretty new, but it is possible. You need a tsup config file for each React package, then add your monorepo packages into your next js app's config file. Also make sure that components shared between your Vite projects and your next js app are designed in a way that they can work with both build systems.

**Reply by heloineto** [reply]
So the config file should tell tsup to build the components in separate files, for the Client Components (with "use client") and Server Components to work correctly?

Also, do you know of any tsup config file of a package that's build with Server/Clients Components in mind? It would really help to have an example.

**Reply by heloineto** [reply]
Also, I don't mind using another bundler.

If I can't figure this out I'm thinking of using vanilla js with no build step

**Reply by sheeehy** [reply, marked as answer]
Yeah, when using tsup with React components that have 'use client' directives for client specific code, you need to configure tsup to build the components into separate files, it basically ensures that the client and server components work correctly. The tsup config file should look something like this, 

```
export default {
  entryPoints: ['src/index.ts'], //  switch to your component's entry point
  format: 'cjs',
  splitting: true, //  split into separate files
};
```
This should generate separate output files for each module. As for examples, I cant think of any that also use server / client components.

**Reply by denu5** [reply]
@sheeehy  thanks for the hint about splitting.

The turborepo boilerplate I found good to start with had splitting set to false 

https://github.com/mkosir/frontend-monorepo-boilerplate/blob/main/packages/ui-mui/tsup.config.ts#L8
